liquidation-monitor/src/agent.js

```diff
@@ -0,0 +1,404 @@
+    if (data.borrow[tokenAddress] === undefined) data.borrow[tokenAddress] = {};
+    if (data.supply[tokenAddress] === undefined) data.supply[tokenAddress] = {};
+    /* eslint-enable no-param-reassign */
+  }
+}
+// #endregion
+
+// Initializes data required for handler
+function provideInitialize(data) {
+  return async function initialize() {
+    // #region Assign configurable fields
+    /* eslint-disable no-param-reassign */
+    data.alertMinimumIntervalSeconds = config.alertMinimumIntervalSeconds;
+    data.protocolName = config.protocolName;
+    data.protocolAbbreviation = config.protocolAbbreviation;
+    data.developerAbbreviation = config.developerAbbreviation;
+    data.alert = config.liquidationMonitor.alert;
+    data.minimumLiquidationInUSD = config.liquidationMonitor.triggerLevels.minimumLiquidationInUSD;
+    data.lowHealthThreshold = config.liquidationMonitor.triggerLevels.lowHealthThreshold;
+    data.cTokenABI = getAbi('cErc20.json');
+    data.provider = getEthersProvider();
+    data.accounts = {}; // Health of all accounts, calcHealth, lastUpdated, [assetsIn addresses]?
+    data.supply = {}; // qty of cTokens (not Tokens)
+    data.borrow = {}; // qty of Tokens (not cTokens)
+    data.tokens = {}; // each cToken's address, symbol, contract, ratio, price, lastUpdatePrice
+    data.newAccounts = []; // New account from transaction events
+    data.totalNewAccounts = 0;
+
+    // Compound API filter and Comptroller contract
+    const {
+      comptrollerAddress,
+      maxTrackedAccounts,
+      oracleAddress,
+      oneInchAddress,
+      triggerLevels: {
+        maximumHealth,
+        minimumBorrowInETH,
+      },
+    } = config.liquidationMonitor;
+    const comptrollerABI = getAbi(config.liquidationMonitor.comptrollerABI);
+    const oracleABI = getAbi(config.liquidationMonitor.oracleABI);
+    const oneInchABI = getAbi(config.liquidationMonitor.oneInchABI);
+    data.comptrollerAddress = comptrollerAddress;
+    data.comptrollerContract = new ethers.Contract(
+      comptrollerAddress,
+      comptrollerABI,
+      data.provider,
+    );
+    data.oneInchContract = new ethers.Contract(
+      oneInchAddress,
+      oneInchABI,
+      data.provider,
+    );
+    // Unused but available for the future
+    data.oracleContract = new ethers.Contract(
+      oracleAddress,
+      oracleABI,
+      data.provider,
+    );
+    /* eslint-enable no-param-reassign */
+    // #endregion
+
+    // #region Get initial accounts from Compound API
+    // Find total number of results with the first request
+    const initialRequest = buildJsonRequest(maximumHealth, minimumBorrowInETH, 1, 1);
+    const initialResults = await callCompoundAPI(initialRequest);
+    const totalEntries = initialResults.pagination_summary.total_entries;
+
+    // Determine number of pages needed to query. Results vs config limit.
+    const maxEntries = Math.min(maxTrackedAccounts, totalEntries);
+    const maxPages = Math.ceil(maxEntries / 100);
+
```

Comment:
This is potentially messy. Depending on how they've implemented result ordering, (if they sort newest -> oldest, so new transactions show up in earlier pages and push other data back) you can potentially get duplicate data, maxPages/entries might be stale by the time you've walked the result, etc

Not sure if we can really do anything, it's an API problem, but it's worth mentioning. Results from this may wind up looking a bit strange with high transaction count traffic.